Stats Widget: Load stats in the widget after user logs in

staskus opened the pull request:
## Issue

Closes #17567 

## Description

### Issue

Data on the stats widget is not shown until the user manually visits the stats page.

### Rootcause

There are 2 reasons:
1. App only calls stats widget initialization in `StatsViewController.initStats()`
2. Stat widget itself in `SiteListProvider.getTimeline(...)` doesn't fetch stats while `configuration.site` is nil

### Solution

1. Listen to `WPSigninDidFinishNotification` and initialize stats when the event is received. 
Two important pieces of data are already loaded when this event triggers: 1) `WPStatsHomeWidgetsUserDefaultsSiteIdKey` is set 2) `visibleBlogsForWPComAccounts` are set

2. Update `SiteListProvider` so it would use `defaultSiteID` while `configuration.site` is nil. This allows Stats widget to load required data itself in `SiteListProvider.getTimeline(...)` and not rely on `StatsViewController` being opened for data to be loaded

### Another potential improvement

Create an empty state for the widget while the data is not loaded. For example show the site name and "-" instead of numbers.

## Testing instructions

1st scenario:

1. Open the app and make sure it's logged out
2. Add Widgets
4. Log in
5. Background the app
6. Confirm that Stats load

2nd scenario:

1. Open the app and make sure it's logged out
4. Log in
5. Background the app
7. Add Widgets
8. Confirm that Stats load

3rd scenario:
1. Open the app and make sure it's logged out
2. Log in with the account
3. Add widget
4. Long tap on the widget and click Edit Widget
5. Select a site
6. Come back to the app
7. Logout
8. Log in with the different account
9. Background the app
10. Confirm the Stats load

## Regression Notes

1. Potential unintended areas of impact

Make sure the widget is still updated when visiting Stats view

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

Added `StatsWidgetsStore` to confirm that Stats widget data is loaded or deleted depending on notifications posted. More thorough tests would require a refactoring of `StatsWidgetsStore` to inject WidgetCenter, AccountHelper, HomeWidgetCache, and other dependencies.

## Video

Add widgets before logging in:

https://user-images.githubusercontent.com/4062343/174248260-740ce27d-a9e6-4aa6-a114-60e1b56f048d.mov

Add widgets after logging in:

https://user-images.githubusercontent.com/4062343/173892420-e5ded3ce-09d3-43c3-a60c-c4fdb2b887f2.mov

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Remove accidentally restored code in StatsInsightStore

This code was moved to StatsWidgetStore in https://github.com/wordpress-mobile/WordPress-iOS/pull/15838/files but added again in
https://github.com/wordpress-mobile/WordPress-iOS/pull/15887/commits/03e8990901881c9b735350d16b53a12ee91e5368

staskus committed:
Initialize Stats widgets when sign in finishes

When WPSigninDidFinishNotification is posted, both defaultSiteId and  blog data is loaded can be used to initialize Stats widgets.

This functionality allows Stats widgets to preload as soon as possible in the app flow.

staskus committed:
Testing StatsWidget data creation and deletion

- Moving BlogServiceMock to a separate file so it could be reused
- Injecting BlogService to StatsWidgetStore so it could be mocked

staskus committed:
Fetch Stats widget data for the default site on initial load

- getTimeline is called on reloadTimelines() call in the app
- configuration.site can be nil until user edits the widget and site list is loaded
- fetch stats for default site so widget shows data as soon as possible

staskus committed:
Revert a comment in StatsWidgetStore

staskus commented:
Hey @Gio2018, I saw you worked on this Widget code before. Maybe you could take a look at this PR? 👀

CC: @guarani

Gio2018 commented on WordPress/Classes/Stores/StatsWidgetsStore.swift:
```diff
@@ -264,6 +265,10 @@ extension StatsWidgetsStore {
 // MARK: - Login/Logout notifications
 private extension StatsWidgetsStore {
 
+    /**
+     When WPAccountDefaultWordPressComAccountChanged is posted after login
+     the site data is not yet loaded and cannot be cached for widgets.
+     */
```
any chance we could use modern-style documentation for methods, using the typical format, like so?

```
/// Store stats in the widget cache
    /// - Parameters:
    ///   - widgetType: concrete type of the widget
    ///   - stats: stats to be stored
```

Gio2018 commented on WordPress/Classes/Stores/StatsWidgetsStore.swift:
```diff
@@ -286,6 +291,22 @@ private extension StatsWidgetsStore {
         }
     }
+
+    /**
+     When WPSigninDidFinishNotification is posted after login
+     the site data data is loaded and widget data can be cached.
+     */
```
same observation for documentation here

Gio2018 commented on WordPress/Classes/Stores/StatsInsightsStore.swift:
```diff
@@ -1205,105 +1205,3 @@ private extension InsightStoreState {
 }
-
-// MARK: - iOS 14 Widgets Data
-private extension InsightStoreState {
```
good catch on spotting this leftover!

Gio2018 reviewed (changes requested):
Hello @staskus ! 

I left some minor code comments, but other than that, I was testing the first scenario and noticed that, once logged in, the widget still does not display stats until I visit the Stats page. Once I do, it does, but the title is set to a different site (This might have something to do with some weird caching that Widgets do under the hood)
Am I missing something? Here's a brief animation of what I saw:

<p align=center>
<img height=500 width=250 src=https://user-images.githubusercontent.com/34376330/173907809-b6431439-2613-4193-8280-ea5c44f896d4.gif>
</p>

staskus commented:
Thanks for the comments and the review! @Gio2018 

There are 2 cases, the first one is not expected, and the second one is:

> I was testing the first scenario and noticed that, once logged in, the widget still does not display stats until I visit the Stats page.
>

1. Yes, you're not missing anything, you shouldn't need to visit "Stats" page for Widget to load. It could be related to caching as you mentioned. After login `reloadTimelines` should trigger the reload of Widgets. The fact that "Log in" error is shown suggests that not even "No data" state is loaded once Widget knows that the user is logged in.

I'll try to understand this issue better.

> Once I do, it does, but the title is set to a different site.
>

2. The site is always set to a default site unless it's changed in the "Edit Widget" menu. It allows having multiple widgets for different sites at the same time.

<img width="317" alt="image" src="https://user-images.githubusercontent.com/4062343/174065542-1218a756-ea5d-4059-a655-3657aede987c.png">

staskus committed:
Ignoring invalid cached selected site configuration

If SelectSiteIntent configuration is selected for one account and then user logs in with different account, Configuration.site remains loaded and needs to be ignored.

staskus committed:
Use Xcode generated documentation format

staskus commented:
@Gio2018 

I identified the case when similar loading issue happens if:
1. Log in with account A
2. Configure widgets
3. Log in with account B
4. Widget doesn't load

Could it explain what happened in your testing case?

I found the issue with cached information of `SelectSiteIntent` configuration. If site is selected in the widget and the user logs out, `Configuration.site.identifier` still exists. [I haven't found a way to remove this information from code.](https://stackoverflow.com/questions/63764083/how-to-reset-intent-extension-configurations-in-widgetkit)

I improved `getTimelines` method to handle this case and added comments to explain such caching case. 

## Testing

Added Testing scenario 3 to the description

## Videos

Before the change:

https://user-images.githubusercontent.com/4062343/174075611-54c26bf2-6190-4ff4-9566-d7e92e70358d.mp4

After the change:

https://user-images.githubusercontent.com/4062343/174075675-ce647087-affa-4565-a18e-28d49021a3e2.mp4

Gio2018 reviewed (approved):
Awesome @staskus , it worked this time! Tested both scenarios.
There is a bit of delay to load stats in the widget after login, meaning that when I log in and background the app, I see "Unable to load stats" for a moment, and then the widget is populated.
Not sure if this is normal or there is something that could be improved in the way we load data (I was testing this on a fast connection so I would expect that data would load fast), but this is certainly something that can be tackled in a separate PR.

staskus committed:
Reload widget timelines only once during login

- WPAccountDefaultWordPressComAccountChanged gets called during login and after logout
- WPSigninDidFinishNotification gets called after login

Only reload timelines after WPAccountDefaultWordPressComAccountChanged notification when user has logged out

staskus commented:
> Awesome @staskus , it worked this time! Tested both scenarios.
> There is a bit of delay to load stats in the widget after login, meaning that when I log in and background the app, I see "Unable to load stats" for a moment, and then the widget is populated.
> Not sure if this is normal or there is something that could be improved in the way we load data (I was testing this on a fast connection so I would expect that data would load fast), but this is certainly something that can be tackled in a separate PR.

Yes, thanks for the review again, @Gio2018! 

I had a wrong impression this delay happens due to `fetchStats` call. However, it actually happens due to excessive calls to `reloadTimelines` after login:
- We call it once after `WPAccountDefaultWordPressComAccountChanged`
- And almost immediately we call it after `WPSigninDidFinishNotification`

What we need to do is:
- `reloadTimelines` after `WPAccountDefaultWordPressComAccountChanged` when user logged out
- `reloadTimelines` after `WPSigninDidFinishNotification` when user logged in

It solved delay issue, the UX is much better now:

https://user-images.githubusercontent.com/4062343/174248260-740ce27d-a9e6-4aa6-a114-60e1b56f048d.mov

staskus committed:
Set WPStatsHomeWidgetsUserDefaultsLoggedInKey before reloading timelines

staskus committed:
Merge remote-tracking branch 'upstream/trunk' into issue/17567-stats-widget-not-working-until-user-visits-stats-screen-in-app

Commit:
update comment

guarani commented:
Thanks for the PR, @staskus! I'm merging this now given that the commits added after the last PR review approval have been discussed and informally approved.

cc @Gio2018
